### ISSUE TITLE
Proposal for merging array or object options in tsconfig file

### ISSUE THREAD

**Issue by sheetalkamat** [microsoft/TypeScript#57486]
### 🔍 Search Terms

"extends", "merge", "compilerOptions", "paths", "outdir", "include", "exclude", "typeRoots", "tsconfig", "ability to merge tsconfig"

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

oday when you extend a config the options written in the final config override the options from the base config. Here is sample of what happens today:

```jsonc
// @fileName: /temp/test/base.tsconfig.json
{
    "compilerOptions": {
        "paths": {
            "@libs/*": ["libs/*"],
            "@project1/*": ["project1/*"]
        },
    }
}
```

```jsonc
// @fileName: /temp/test/project1/tsconfig.json
{
    "extends": "../base.tsconfig.json",
    "compilerOptions": {
        "paths": {
            "@environment/*": ["env/*"]
        },
    }
}
```

Here is what it would mean that project1 tsconfig was written as:

```jsonc
// @fileName: /temp/test/project1/computed.tsconfig.json
{
    "compilerOptions": {
        "paths": {
            "@environment/*": ["env/*"]
        },
    }
}
```

Over years users have asked us for ability to create a final config that adds to base config instead of overriding. This is extremely useful for `paths` from compiler options.

This would allow to users to write a base config that contains common options that shouldn't have to be copied in all projects and yet allowing projects to specify there additional specific options.

For example in above sample, one would expect a way to write config such that final `paths` property is:

```jsonc
// @fileName: /temp/test/project1/computed.tsconfig.json
{
    "compilerOptions": {
        "paths": {
            "@libs/*": ["../libs/*"],
            "@project1/*": ["../project1/*"],
            "@environment/*": ["env/*"]
        }
    }
}
```

Here are different proposals that are being considered:

1. Add `merge` as root level config property that specifies options that need to merged instead of completely overwritten in the final config.
From the above example writing the `project1` config as:
```jsonc
// @fileName: /temp/test/project1/tsconfig.json
{
    "extends": "../base.tsconfig.json",
    "compilerOptions": {
        "paths": {
            "@environment/*": ["env/*"]
        },
    },
    "merge": ["paths"]
}
```

2. Inline `merge` options:
```jsonc
// @fileName: /temp/test/project1/tsconfig.json
{
    "extends": "../base.tsconfig.json",
    "compilerOptions": {
        "merge": {
            // All the options that need to be merged
            "paths": {
               "@environment/*": ["env/*"]
            },
        },
    },
}
```

3. Specifying options that need to be merged as `merge:optionName` or `merge-optionName`
```jsonc
// @fileName: /temp/test/project1/tsconfig.json
{
    "extends": "../base.tsconfig.json",
    "compilerOptions": {
        "merge:paths": {
            "@environment/*": ["env/*"]
        },
    },
}
```

4. Add root level `merge: true` that merges all arrays and objects options:
From the above example writing the `project1` config as:
```jsonc
// @fileName: /temp/test/project1/tsconfig.json
{
    "extends": "../base.tsconfig.json",
    "merge": true,
    "compilerOptions": {
        "paths": {
            "@environment/*": ["env/*"]
        },
    },
}
```

TODO:
[ ] Are there order dependency risks
[ ] List all the options that are array or object and see if merge makes sense for them and decide if we want to allow only certain options and not all array or object type of options

### 📃 Motivating Example

A way to specify say project specific "paths" mapping without having to rewrite all the mappings from the base config.

### 💻 Use Cases

1. What do you want to use this for?
2. What shortcomings exist with current approaches?
3. What workarounds are you using in the meantime?

**Comment by jxdp** [comment]
Can you provide an example of a project structure that this would actually be useful for?

I am having a difficult time thinking of any that isn't extremely convoluted by design (and I don't those are a problem for TypeScript to solve).

The beauty of `tsconfig` is that it is pretty much WYSIWYG. Every key will either have the value I specify or the value in the config I extend from with no ambiguity.

**Comment by RyanCavanaugh** [comment]
OP for this one is #44589 where there are many people complaining that it's not possible. Probably ask them 🙂

**Comment by SainteCroquette** [comment]
I think main motivation would be **mono-repositories**, where you would want to share typescript configuration across you repository's projects, which most of the time include paths.

Some real world example with [Nx](https://nx.dev/), a popular monorepo management solution  :

[nx integrated monorepos](https://nx.dev/getting-started/tutorials/integrated-repo-tutorial) typically have a root level `tsconfig.base.json` which is then extended by every app / library / package in the repository from which theses modules will find paths definitions among others:
*(generated with [nx's cli](https://nx.dev/nx-api/js/generators/library))*

**./tsconfig.base.json**
```typescript
// define libraries path so that they can be used in other local packages. 
{
  // ...
  "compilerOptions": {
    // ...
    "baseUrl": ".",
    "paths": {
      "@myorg/some-lib": ["libs/some-lib/src/index.ts"],
      "@myorg/some-other-lib": ["libs/some-other-lib/src/index.ts"],
    }
  },
}
```
**./apps/some-app/tsconfig.json**
```typescript
// project specific options 
{
  "extends'": "../../tsconfig.base.json",
  // ...
  "compilerOptions": {
  // theses are local path aliases, only known to this module. needs full path as baseUrl is set in root tsconfig.base.json
    "paths": {
      "@somelocalalias/*": ["apps/some-app/src/somelocalfolder/*"],
    }
  },
}
```
Current behavior makes maintaining local imports a pain :/
Workarounds do exists but most of the time they involve custom scripts or adding a new dependency which is far from ideal for such uses cases as there is no opinionated way of achieving this.

**Comment by jxdp** [comment]
@SainteCroquette Still not really understanding.  This is the exact problem that a monorepo solves--using packages.

In other words, why would you use inherited tsconfig paths instead of dependencies?

```json
{
  "dependencies": {
    "@myorg/some-lib": "*",
    "@myorg/some-other-lib": "*"
  }
}
```

```ts
// my-app/some/file.ts

import { something } from "@myorg/some-lib";
import { somethingElse } from "@myorg/some-other-lib";
```

**Comment by SainteCroquette** [comment]
I can't speak for other tools but if you follow `Nx` (which doesnt rely on package name to resolve paths) [integrated monorepository](https://nx.dev/getting-started/tutorials/integrated-repo-tutorial) guidelines (which is really easy thanks to their cli & plugins) you end up with:
- A root `package.json` listing all the external dependencies (dev included) used by all local repository packages (meaning you enforce sharing the same version of a dependency for all your consumer packages).
- `package.json` for your packages without a single dependency listed (everything from root `package.json` is available to you anywhere, but you can still list somemore here if you want)
- A root `tsconfig.base.json` which the Nx cli will update when addind / moving / removing local package from your repository.


Inside a package, build tools such as webpack or vite can then use a plugin provided by nx to feed them the path options from the local tsconfig. To be able to map correctly to other packages from your mono repository the local tsconfig needs to extends base tsconfig which holds the mapping configuration of the whole repository.
Defining a path option for the local tsconfig would then break the whole toolchain.
Would this means that nx's choice of relying on tsconfig as 'single source of truth' is against intended usage of tsconfig's path options?


Could you expand a bit on sharing tsconfig files via dependencies and how this would allows you to merge to path configurations together (or is this another way to achieve the same thing ?) ?

**Comment by jxdp** [comment]
> Could you expand a bit on sharing tsconfig files via dependencies and how this would allows you to merge to path configurations together (or is this another way to achieve the same thing ?) ?

@SainteCroquette  [Package-based monorepos (ie npm workspaces) also work with Nx](https://nx.dev/concepts/integrated-vs-package-based#packagebased-repos). The Nx integrated plugin "magic" is completely opt-in.  In other words, the need for merging tsconfig paths in such a setup is a direct consequence of the way Nx internals work. As such, it seems more reasonable to me that Nx (in this case) should provide a solution (eg via another plugin).

The question is, where does TypeScript draw the line about what it supports out of the box?  To me, it feels right that it works with the standard npm/node resolution.  Anything beyond that, I think, it likely to lead to endless scope-creep.

**Comment by Bessonov** [comment]
Just to mention that Webpack uses special `'...'` syntax to [include defaults](https://webpack.js.org/configuration/optimization/#root:~:text=Basically%2C%20%27...%27%20is%20a%20shortcut%20to%20access%20the%20default%20configuration%20value%20webpack%20would%20otherwise%20set%20for%20us.).

**Comment by saltman424** [comment]
@jxdp one issue with TypeScript choosing to rely on npm workspaces/dependencies as the primary (or only) supported setup for mono-repos is that it unnecessarily couples TypeScript to node/npm for mono-repos. There are other runtimes for TypeScript, such as [Deno](https://deno.com/) and [Bun](https://bun.sh/). Personally, I prefer node, but it seems to me that a self-contained TypeScript solution for mono-repos has some advantages.

Now you could take your point more generically, that mono-repos are outside the scope of TypeScript. I.e. the burden of making TypeScript useful in mono-repos is entirely on npm, Nx, or whoever else wants to take it on. One problem with that is TypeScript is already part of the way there. Just using `extends` and `paths` allows you to easily create buildable code in a mono-repo. So at best, there is going to be a split burden because TypeScript mono-repos, whether or not they are using a third-party tool like npm or Nx, are sometimes going to use or these properties but then are also limited by them, and thus we will likely continue to see interest in issues like #44589.

As for the question on use cases, I think the Nx integrated mono-repo approach that @SainteCroquette mentioned is a reasonably common one. And I disagree with
> the need for merging tsconfig paths in such a setup is a direct consequence of the way Nx internals work

At least in my use case, I use Nx just to orchestrate tasks. My code is able to build and run entirely without Nx. We could easily swap Nx out for a different task orchestration framework and still build and run all the same TypeScript code with all the same tsconfigs. So it would be nice to not have to copy-and-paste paths between those tsconfigs.

In fact, Nx effectively does not intrude on the actual mechanics of sharing code in your project by allowing you to either use npm workspaces (package-based) or TypeScript paths (integrated). In other words, they seem to be treating it as if npm and TypeScript provide two different, valid ways for sharing code in a mono-repo, and they just try to support both.

Also, to this objection:
> The beauty of tsconfig is that it is pretty much WYSIWYG. Every key will either have the value I specify or the value in the config I extend from with no ambiguity.

I would argue that the `extends` property makes the tsconfig precisely not WYSIWYG. Especially, since you can have an indefinite chain of `extends`. It is not uncommon for me to have to go through a series of extended tsconfigs to find the property I am looking for, only to find it is not specified anywhere, at which point I have to look up the default value. Also, I don't think most of the proposed changes here would meaningfully alter the degree to which tsconfig is WYSIWYG. You would still see what is specified in your tsconfig and then have to look at the extended tsconfig for anything where your tsconfig indicates that it is using the extended tsconfig (i.e. by omitting the property or by explicitly merging it). So as long as a relatively intuitive approach is used for merging, not much changes. And I think the point of this issue is to evaluate how intuitive and useful each of the proposed approaches are.

## TLDR

1. With the `extends` and `paths` properties of tsconfigs, it is not hard to create buildable code in a TypeScript mono-repo without any other tooling
2. This merging feature would make that experience even better, especially for `paths`, and regardless of what other tooling is used
3. tsconfigs are already **not** WYSIWYG

**Comment by SainteCroquette** [comment]
@saltman424 while templates are powerful & flexible and they do solve current problem, they often lacks readability. They might be too much *vodoo* for that kind of use case. 

Personally, I favor `comprehensiveness` over `conciseness`. Moreover that would probably increase overall complexity of tsconfig files, going against its straightforward declarative syntax.

Proposals **1**, **2** and **3** are explicit in that sense.
Not sure how you would handle the case where there is multiple files are extended tho #57567

**Comment by saltman424** [comment]
@SainteCroquette that's fair. I mostly just don't want TypeScript to have to continually release refinements and new properties to address every edge case with merging. The main appeal to me with template variables is it gives me a lot of control so if there are edge cases that come up, I can resolve them myself. And the results are just more predictable. I find behind-the-scenes magic great when it works and really frustrating when it doesn't.

With that being said, the default merge logic would probably work for most of my current use cases, so I am not opposed to it. For multiple extends, I would assume the merge logic would just be applied in order of extends. I.e.
```js
{
    "extends": [ "./a", "./b"],
    "compilerOptions": {
        "paths": {
            "@mystuff/*": ["./*"]
        },
    },
    "merge": [ "paths" ]
}
```
would use `paths = merge(merge(a, b), { "@mystuff/*": ["./*"] })`

And now that I think about it, another potential benefit of merge properties would be it encourages developers to split their extended tsconfigs into isolated configurations. I.e. with template variables, you could selectively pull out the parts of a tsconfig that you want, while with merge properties, you are merging an entire property from all of your extended tsconfigs, so you need to make sure you only have exactly what you need in the extended tsconfigs.

Side note: I would think template variables would be more `comprehensive` (cover more use cases more completely), while merge properties would be more `concise` (require less input from the developer)

**Comment by jxdp** [comment]
@saltman424 

> Now you could take your point more generically, that mono-repos are outside the scope of TypeScript. I.e. the burden of making TypeScript useful in mono-repos is entirely on npm, Nx, or whoever else wants to take it on. 

Yes, I would generally lean towards that as a guiding principle.

> In fact, Nx effectively does not intrude on the actual mechanics of sharing code in your project by allowing you to either use npm workspaces (package-based) or TypeScript paths (integrated). In other words, they seem to be treating it as if npm and TypeScript provide two different, valid ways for sharing code in a mono-repo, and they just try to support both.

But npm workspaces and TypeScript paths are not fundamentally not equivalent.  npm workspaces are a dependency management solution; TypeScript is a DSL and a transpiler.

I am very skeptical about TypeScript adding complexity to handle problems that are complex, optional and can be resolved in user-land without any changes to TypeScript.

In the Nx case:
- dependency resolution is complex
- using tsconfig paths for dependency management is optional
- it could be resolved fairly simply by letting Nx generate the tsconfig files (it already has Nx specific config files and understands your entire dependency tree, so this should be a fairly straightforward plugin–I wouldn't be surprised if it already exists)

**Comment by saltman424** [comment]
@jxdp

> But npm workspaces and TypeScript paths are not fundamentally not equivalent. npm workspaces are a dependency management solution; TypeScript is a DSL and a transpiler.

I totally agree. TypeScript paths are not meant to manage dependencies. They make that quite explicit in the documentation. However, it is a valid way to tell TypeScript that you are using another tool to resolve a given path, so you would like TypeScript to understand that path as well. When working with a bundler or executor, this means you can split your code into distinct folders without having them be separate packages, but instead all as input folders for bundling (i.e. every folder will be subject to the same compiler options; have the same dependencies available to them; and be built/transpiled together).

So I agree, these are not interchangeable solutions. They are two different approaches to mono-repos. But they are both viable approaches.

With that being said, this does make me realize that it is not Nx, but really bundlers, like esbuild, and executors, like tsx, that makes this approach to mono-repos viable. The problem, I would argue, is it definitely is not the scope of bundlers or executors to create/manage tsconfigs. They simply consume the tsconfigs for bundling or executing. So I think TypeScript is the most natural place to add this functionality, as numerous bundlers and executors seem to already rely on tsconfigs for specifying this type of path resolution. However, I hear the argument that it was not really the original purpose of `paths` to become this sort of configuration option for third-party tooling. I just think whether it was intentional or not, this is where we are, so why not make it a little more user friendly. And regardless, `paths` is not the only property that could benefit from merging and mono-repos aren't even the only case where merging would be helpful.

> I said *pretty much* WYSIWYG. Every property is either the value in your tsconfig, or the value in the config it extends from. Even in a "Russian doll" scenario, you still end up on a single property (or none)–you do not need to mentally merge all the layers in your head to understand what your actual config is.

I think this depends on the person. I find the hopping around much more burdensome than having to append two lists or merge two objects in my head, but others may experience the reverse.

I also see a fair amount of confusion from a lack of merge functionality. E.g. developers confused why including a folder excludes another folder (because `include` overwrites the extended `include`).

Also, with template variables, it is all explicit, so I don't think that adds a lot of cognitive complexity, at least for a reader. And we already have merge logic on `compilerOptions`, so I think even with magic merge properties, people seem to be at least somewhat capable of understanding how it works without too much mental burden.

**Comment by mrdeveloper-ir** [comment]
is there any fix here ?

**Comment by silverwind** [comment]
Imho typescript would be better off just supporting `typescript.config.{js,ts}`, then you can offload all this merge/extend/etc logic to the user doing it in their code. Numerous tools already support `.config.{js,ts}` files.

**Comment by hichemfantar** [comment]
My suggestion is to change the current behavior of to extending `compilerOptions` by default instead of overriding them.
And I suggest a new propertry "override" that allows setting which properties will be overriden.

```json
{
  "extends": "@acme/tsconfig",
  "compilerOptions": {
    // this will override all the compiler options from the base tsconfig
    "override": ["types"],
    "baseUrl": ".",
    "types": [
      "node",
      "@acme/module-type-aliases",
      "@acme/theme-classic",
      "acme-plugin-sass"
    ]
  },
  "exclude": ["node_modules"]
}
```

**Comment by PatrykKuniczak** [comment]
This is bizzare, in 2024, this feature isn't implemented.

I want to use it for `jsx, baseUrl, types` but i can't.

This should work like [deepmerge](https://www.npmjs.com/package/deepmerge#mergeallarrayofobjects-options)

**Comment by ciandt-crodrigues** [comment]
Is there any workaround for this to reference the extended config on the project paths?

**Comment by ciandt-crodrigues** [comment]
To be honest, having a tsconfig.ts where we could freely compose all the strong type configuration file fields (even dynamically if we want) would be much better than these merge options

**Comment by dkamins** [comment]
Of the proposed "merge" named solutions, I prefer the merge-modified key names (3,4) over extra merge meta keys (1,2,5), and would say definitely not 8 which would be ambiguous/confusing. Overall though, my top choice would be a twist on 6 ("Simple placeholder like ${base}: true or <merge>: true").

As suggested by @Bessonov (above in https://github.com/microsoft/TypeScript/issues/57486#issuecomment-1966259430), as is used [in webpack](https://webpack.js.org/configuration/optimization/#root:~:text=Basically%2C%20%27...%27%20is%20a%20shortcut%20to%20access%20the%20default%20configuration%20value%20webpack%20would%20otherwise%20set%20for%20us.), similar to [GraphQL fragments](https://graphql.org/learn/queries/#fragments), approximating what we are familiar with in TS/JS with the [`...` spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), and just generally familiar, how about using `...` as the extension placeholder?

Benefits:
- Immediately visually obvious what it is doing.
- Easy to remember (vs "_I forgot, is it key:merge or merge:key?_")
- Lists can be easily prepended, appended, and even both at once.
- Supports simple meta-tooling without requiring encoding or scanning additional key names, parsing key names, encoding of merge logic. (I.e. tools that only process or view tsconfigs can just pass the spreads through directly as the strings they are.)
- Backward compatible with earlier tool versions for the same reason.

```ts
// @fileName: /temp/test/project1/tsconfig.json
{
    "extends": "../base.tsconfig.json",
    "compilerOptions": {
        "paths": {
            "...": "...",                                // extends!
            "@environment/*": ["env/*"]
        },
        "typeRoots": ["mytypes1", "...", "moretypings"]  // extends!
    },
}
```

I chased this issue trying to extend path aliases 
- from https://github.com/microsoft/TypeScript/issues/44589 (2021, still open)
- to https://github.com/microsoft/TypeScript/issues/56436 (2023, closed)
- finally to here. 

Now I'm feeling good about 2025 being the year we can all extend our tsconfigs into the future together! 🚀

💪`...`👍

**Comment by ijxy** [comment]
@dkamins I don't think that a magic "..." string is a very compelling option.

Given that you have very specific requirements, I suggest that you _generate_ your TSConfig files using whatever programming language you prefer.

Pros:
- the TSConfig spec, which works great for what it is intended, remains unchanged
- you have complete flexibility to do whatever crazy things you want, with whatever syntax you want
- you could get support from likes of `nx` who seem to lean towards using TSConfig paths instead of proper dependency management

Cons:
- you have to do the work

**Comment by PatrykKuniczak** [comment]
I think it'll never be implemented, because the clock is ticking and nothing special happend here.

**Comment by SainteCroquette** [comment]
Well yeah I'm not sure we could find the perfect solution on first try.
A rather naïve implementation as an experimental feature might be the way to go. Then refine if it is needed.

So naïvely :
- Everything that can live along each other get merged together
- Everything that is exclusive of each other get overwritten by the latest value in the chain
- Path are relative to were they are defined

e.g

from base
```json
// @fileName: base.tsconfig.json
{
    "compilerOptions": {
        "paths": {
            "@libs/*": ["libs/*"],
            "@project1/*": ["project1/*"]
        },
        "typeRoots": ["node_modules/@types"],
        "skipLibCheck": true
    }
}
```
extended by
```json
// @fileName: project1/tsconfig.json
{
    "extends": ["../base.tsconfig.json"]
    "compilerOptions": {
        "paths": {
            "@some/*": ["something/*"],
        },
        "typeRoots": ["node_modules/@types"],
        "skipLibCheck": false
    }
}
```
would output

```json
// @fileName: project1/tsconfig.json
{
    "compilerOptions": {
        "paths": {
            "@some/*": ["something/*"],
            "@libs/*": ["../libs/*"],
            "@project1/*": ["../project1/*"]
        },
        "typeRoots": ["node_modules/@types"],
        "skipLibCheck": false
    }
}
```

**Comment by 610** [comment]
Hey there! I understand how challenging this situation has been, and I’ve been monitoring it for about a year. It’s a little disappointing that the MS TS Core team ignores our requests, which seems very logical to most people. I’d love to share my 2 cents on this, hoping it might save some time for some people!

 **I apologize if this post is unreasonably long.** :wink:

---
> [!NOTE]
> My information may be outdated because I stopped working with **Nx** a year ago. Please correct me if you, the reader, know the solution to the problem.

### Problem with Nx monorepo tool.

I faced this issue while implementing a monorepo. The most frustrating aspect was **Nx's** requirement to define aliases only for shared libraries (aka **libs** in `/libs/*` directory) in the `tsconfig.base.json` (e.g., `"@my-large-project/router": ["libs/router/src/index.ts"]`). This means that sub-projects (aka **apps** in the `/apps/*` directory) can import shared libs using path aliases, but all imports related to an app have to be either absolute or relative paths (no aliases allowed, no way to add). <ins>You can imagine the number of lines and rows import statements take in the `.ts` files.</ins>
I spent many hours trying to extend the `paths` property in the app's `tsconfig.json` file. Unfortunately, I was unsuccessful because the app's tsconfig file took precedence over the base config, removed libs aliases, and the **Nx** build began to fail.
I read numerous threads on this topic and came across official comments from MS TypeScript core developers and contributors posted a simple response (**<ins>Warning:</ins>** not an exact quote):

> The `paths` property is a complex array type, so it is technically impossible to merge them.

Although I noticed other complex properties within the `compilerOptions` object could be merged, but they were adamant, declined to continue the discussion, and began rejecting related feature requests. 

**Conclusion for Nx.**
If you come here trying to merge `path` or somehow add path aliases for your apps and libs in Nx, - do not waste time. It doesn't allow appending app-related aliases to the `tsconfig.base.json` config.

---
### Custom Monorepo from scratch.
This could be a workaround for projects with not many shared libs and apps.

I avoided using **Nx** for my next big project. I didn't even want to waste hours investigating, testing, and comparing other tools such as **TurboRepo**, **Rush**, **Lerna**, **Moon**, **Bazel**, and *God knows what else that Bull dropped on a barn called **npmjs***. This is mainly because I didn't need most of the features and automation and often had to invest twice as much time on trivial tasks. The really crucial features and helpers can be easily replicated.

**For example,** given a task to create a monorepo for an enterprise project that has multiple backend services or microservices and multiple front-end websites or web-based mobile applications. While backend and front-end projects have different functionality and purposes, they always share some functionality, most often utilities and data models. These can be moved to a shared library and imported where needed. This is a time-proven strategy for maintaining your project's source code clean, consistent, scalable, and fool-prove.

**Here is what I recommend:**
- Use `Yarn` workspaces (https://yarnpkg.com/features/workspaces). You can have package.json in the root for packages you need in most of your apps and libs and only required packages for apps and libs.
  - Run `yarn -W add [package-name]` in the root of your project. These packages will be available in all your apps and libs.
  - Go to particular app or lib directory and run `yarn add [package-name]` as usual. This will install packages that will be available and used only for the app or lib you are currently in.
- You don't need `tsconfig.base.json` in your project root—it is useless anyway. Instead, add individual `tsconfig.json` files for each app and library. Yes, you will have duplicated tsconfig configs, but you update it so rarely that it is barely an issue.
- You can still share `.eslintrc.json`, keep it in the project's root, and extend it to your apps and libs, there are no issues there.
- Depending on your needs, you will need to write custom shell scripts to build the whole project or automate something, but in my case, Github Actions takes all the heavy lifting on building and deploying the project.

:wave: **I may have missed something. Please don't hesitate to ask me any questions. I will try to help and update my post.**

**Comment by JamesHenry** [comment]
@610 I'm sorry to hear about your Nx at that time, **I'm pleased to say that is indeed out of date in very significant ways.**

Nx's TypeScript setup is now fully project references based, and does not use path mappings at all (you could technically continue to leverage them) but after collaboration with the TypeScript we decided to move away from them in favour of package manager workspaces and standard package.json imports/exports for resolution.

That additional bonus is that `tsc -b` from the root of the monorepo now "just works" if you happened to want to type-check all your projects in a single process (naturally not super scalable for large monorepos). This is because Nx is purely a complement to TypeScript, it is not hiding anything away or performing any unpredictable magic.

Please check out all our recent material on this:

**Nx's new approach (very different from what you describe)**

https://nx.dev/blog/new-nx-experience-for-typescript-monorepos

**General guidance and background on the history of TypeScript monorepos and different approaches**

https://nx.dev/blog/typescript-project-references
https://nx.dev/blog/managing-ts-packages-in-monorepos

**Comment by 610** [comment]
@JamesHenry It appears that I have unconsciously concealed my long-standing anger toward **Nx** in the comment of the thread about a little different topic. And I mentioned **Nx's** name so many times that I instantly "_spawned_" you into the conversation. :joy:
Thank you for the update! Your new solution is fantastic. :100: Now, we can consider trying **Nx** again when the time comes.

> P.S. I also apologize if my original comment seemed offensive.

**Comment by JamesHenry** [comment]
@610 Ha no worries at all, I was subscribed to this thread already because of interest in the underlying topic, so I was able to see your response. 

We are really excited to be able to talk about all these updates now because we've been working on them for months and so your message was well timed for me to be able to link all these blog posts! It hasn't landed yet, but we have also managed to built out an automates migration generator/codemod to help with 90% of the transition from the old world to the new for existing Nx workspaces, so folks can stay tuned for that, it will land prior to Nx 21.

We can leave the Nx specific discussion there on this thread so that we don't cause too many extra notifications for the many subscribers on here, but if anyone else wants to discuss Nx feel free to join our active Discord: https://go.nx.dev/community

**Comment by luke-lacroix-healthy** [comment]
Would a simple solution to this be to have Typescript allow for `tsconfig.js` or `tsconfig.ts`? This is how I've handled dynamic configurations when using other tools without complicating the tool itself.
This let's every project decide when/if to merge properties and how to do so. It also removes the need for the `extends` property because you can do that by importing the base configuration(s) and merging them as you want.

**Comment by ijxy** [comment]
> Would a simple solution to this be to have Typescript allow for `tsconfig.js` or `tsconfig.ts`? This is how I've handled dynamic configurations when using other tools without complicating the tool itself. This let's every project decide when/if to merge properties and how to do so. It also removes the need for the `extends` property because you can do that by importing the base configuration(s) and merging them as you want.

This sounds a lot like the new `eslint` config setup. In my opinion, if there is to be some change then this idea makes a lot more sense that adding bespoke JSON merging rules that are specific to TSConfig. Merging would remain a user-land problem, but the user would now have the full power of the JS ecosystem at their disposal.

**Comment by luke-lacroix-healthy** [comment]
@ijxy  I did some digging. It looks like the official position of the maintainers, [as of 2019](https://github.com/microsoft/TypeScript/issues/30400#issuecomment-536350685) by @RyanCavanaugh , is that they would not support other configuration file types and that we should use code that runs as part of our build process to generate the `tsconfig.json` file from whatever sources or methods we choose.

I think this applies to the OPs proposal as well - the complexity of implementing one or more types of merge semantics within a JSON document seems to be the exact use case for either `tsconfig.js` (my preference) -or- an external tool. It would be great if a maintainer could weigh in on this.

**Comment by pcfreak30** [comment]
So I will give my 2 cents on this.

a very common thing I see and do is.

```
"compilerOptions": {
 "paths": {
   "X":"Y",
}
}
```

In a root config. I then would like to re-use these across packages in a mono repo by extending the base *and not need to redefine them*.

I do not use NX or yarn. I use pnpm and do use workspaces. I personally also think just eval'ing a ESM js config file would future proof everything as trying to create a DSL for JSON is going to be error prone.

This is something that also feels common sense to me. The arguments saying `you will never touch them again` defeats the point of the fact this isn't DRY and is unnecessary.

So I would like to see either a way to extend, additive, or to just get JS control so the author can do what they want themselves.
